api/envoy/service/discovery/v3/discovery.proto

```diff
@@ -272,4 +273,13 @@ message Resource {
 
   // The resource being tracked.
   google.protobuf.Any resource = 2;
+
+  // Time-to-live value for the resource. For each resource, a timer is started. The timer is reset each time the
+  // resource is received with a new TTL. If the resource is received with no TTL set, the timer is removed for the
```

Comment:
Where did we land with more efficiently updating the TTL without requiring full wire transmission of the resource which could be large? I thought we had discussed allowing an empty resource, but same version and TTL to update the TTL, kind of like an etag? 

My concern with not doing this now is it may be more confusing to retrofit this later. Is it worth considering doing this optimization now? As a follow up later if needed? Or do we not think it will be needed at all?

Reply:
@mattklein123 The problem is that VHDS uses empty messages already: https://github.com/envoyproxy/envoy/pull/13201#discussion_r504966954.

This means that we'll have to do some work to untangle this bit of tech debt, as otherwise we don't have a way of disambiguating a heartbeat response and a valid VHDS update. Some options are mentioned in the linked thread, but none of them seem amazing to me. 

It would be pretty nice to have these special updates for SotW control planes (e.g. go-control-plane), as special heartbeat responses would allow us to respond with just the TTL'd resources in the response without worrying about whether the resource type is a collection (CDS/LDS) or not.

Reply:
I think we could just add an additional bool that clearly indicates a heartbeat response? I don't love it as it should be a oneof, etc. but I think it would work. If we go this route I guess this could be done later?

Reply:
I think adding a bool is conceptually the easiest thing, though it leaves the API messier than it could be if we could somehow get rid of the VHDS case. Would we be worried about clients that don't know about this field interpreting a heartbeat update as a real update? I guess we could use a client capability to guard this, though I'm not sure if most control plane check this in practice.

@htuch @markdroth in case either of you have more opinions here

Reply:
Rather than adding a bool to the API, I would prefer to just special-cae the VHDS behavior for now, with the understanding that we'll eliminate that special case as part of migrating VHDS from aliases to the new udpa naming scheme.

Reply:
> Rather than adding a bool to the API, I would prefer to just special-cae the VHDS behavior for now, with the understanding that we'll eliminate that special case as part of migrating VHDS from aliases to the new udpa naming scheme. 

This would be fine with me. What do we need to do in this PR to make sure that we can support this in the future, or should we just do it now?

Reply:
I'm taking a look at what it would take to intercept all heartbeat responses except the VHDS ones in the mux impls, if that isn't too bad then I'll just include that in this PR.

Reply:
OK sounds good. Thanks for looking into this. At minimum if we don't implement now add some TODOs?

/wait

Reply:
Alright I pushed the relevant changes with a few tests. 

I realized that the only reason why the VHDS thing is an issue is that the delta code will trigger updates even if the version of the updated resources didn't change. It might be okay to say that if resource empty && version didnt change we won't trigger the subscription callbacks. This would break VHDS implementations that don't increment the version when it sends delta updates, but it might be that those are rare enough that we can get away with it?

Reply:
> This would break VHDS implementations that don't increment the version when it sends delta updates, but it might be that those are rare enough that we can get away with it?

Add a runtime flag and release note and if no one complains we can just remove it later?

Reply:
I think it's a violation of the spec to send new contents without changing the version.  I'd be totally fine breaking any such usage, even without a flag.